10x/wdls/count.wdl

```diff
@@ -0,0 +1,1072 @@
+  File chunks
+  File chemistry_def
+  String barcode_whitelist
+  Int reads_per_file
+  Float subsample_rate
+  Array[Map[String, String]] primers
+
+  File split_file
+  Array[File] fastq_chunks
+ 
+  String d = '$' 
+  String l = '{'
+  String r = '}'
+ 
+  command {
+
+    # Put the fastq file into the $PWD/files directory where martian is expecting
+    # them
+    mkdir files
+    for fastq_chunk in ${sep=" " fastq_chunks}; do
+      ln "${d}fastq_chunk" files/"${d}(basename ${d}fastq_chunk)"
+    done
+    
+    # Read the chunks and chemistry json files to strings so we can pass them as
+    # a Martian map
+    chunks_string=${d}(<${chunks})
+    chemistry_def_string=${d}(<${chemistry_def})
+
+    /opt/tenx/run_in_10x_env.bash martian stage run EXTRACT_READS main \
+      --chunks "${d}chunks_string" \
+      --chemistry_def "${d}chemistry_def_string" \
+      --barcode_whitelist ${barcode_whitelist} \
+      --reads_per_file '${reads_per_file}' \
+      --subsample_rate '${subsample_rate}' \
+      --skip_metrics no \
+      --primers '[${sep="," primers}]' \
+      --split_file '${split_file}'
+    
+    # The Martian cli adapter dumps these outputs as json file, so convert them
+    # to a format Cromwell can parse
+    jq -r '.[]' read_groups > read_groups.lines
+    jq -r '.[]' gem_groups > gem_groups.lines
+    jq -r '.[]' bam_comments > bam_comments.lines
+
+  }
+
+  runtime {
+    docker: "10x_hca_wdl:latest"
+  }
+
+  output {
+    Array[File] trimmed_seqs = glob("trimmed_seqs.bam/*")
+    Array[File] reads = glob("reads.fastq/*")
+    File barcode_counts = "barcode_counts.json"
+    File chunked_reporter = "chunked_reporter.pickle"
+    Array[String] read_groups = read_lines("read_groups.lines")
+    Array[Int] gem_groups = read_lines("gem_groups.lines")
+    Array[String] bam_comments = read_lines("bam_comments.lines")
+    File align = "align"
+  }
+}
+
+task extract_reads_join {
+  String align 
+  String barcode_whitelist
+  File chemistry_def
+  Array[Array[File]] reads_output
+  Array[Array[File]] trimmed_seqs_output
+  Array[Array[String]] bam_comments_output
+  Array[Array[Int]] gem_groups_output
+  Array[Array[String]] read_groups_output
+  Array[File] chunked_reporter_output
+  Array[File] barcode_counts_output
+    
+  Array[File] split_files
+  
+  String d = "$"
+  String l = "{"
+  String r = "}"
+
+  command {
+    set -x  
+    
+    # There's a lot here, but it's mostly the same thing repeated. All the input reads and trimmed_seq
+    # files need to be moved to paths in $PWD. Also, we have to prepare a json list of lists that
+    # has all these new paths so we can pass it to martian. The jq ' |= .+ [ ]' syntax just appends 
+    # something to a list. Finally, we need to create a list of empty lists called read2s that's the
+    # same length as the list of reads.
+    idx=0
+    reads_arr_arg=[]
+    read2s_arr_arg=[]
+    for reads_arr in ${d}(echo '[${sep="," reads_output}]' | jq -c '.[]'); do
```

Comment:
I guess there is something nice and clean doing everything in bash but I sure do like `.append()` more than `jq ' |= .+ [ ]'`

Reply:
I like to think of it as having an austere beauty, but maybe a little writeup of jq tricks would help?

Reply:
Hah. :smile: Could you use a lambda to define it once and then reuse the defined function? That would clarify and reduce code clutter (?)

Reply:
I'm not sure how to do that in bash in a way that doesn't feel very exotic. We could have some named functions, but I think we'd have to define those in every step. We could maybe add some simple executables in the docker image called something like `jq-append`. That would clean up the WDL a lot, though the tradeoff would be that you'd have to inspect the image to see what's really going on. What do you think?

Reply:
I think I prefer the existing method. There's a bit more code but at least it's all there and the user can look it up if they want to know what it's doing
.